oci/private/push.bzl

```diff
@@ -0,0 +1,106 @@
+"Implementation details for the push rule"
+
+_DOC = """Push an oci_image or oci_image_index to a remote registry.
+
+Pushing and tagging are performed sequentially which MAY lead to non-atomic pushes if one the following events occur 
+
+- Remote registry rejects a tag due to various reasons. eg: forbidden characters, existing tags 
+- Remote registry closes the connection during the tagging
+- Local network outages
+
+In order to avoid incomplete pushes oci_push will push the image by its digest and then apply the `default_tags` sequentially at
+remote registry. 
+
+Any failure during pushing or tagging will be reported with non-zero exit code cause remaining steps to be skipped.
+
+Push an oci_image to docker registry with latest tag
+
+```starlark
+oci_image(name = "image")
+
+oci_push(
+    image = ":image",
+    repository = "index.docker.io/<ORG>/image",
+    default_tags = ["latest"]
+)
+```
+
+Push an oci_image_index to github container registry with a semver tag
+
+```starlark
+oci_image(name = "app_linux_arm64")
+
+oci_image(name = "app_linux_amd64")
+
+oci_image(name = "app_windows_amd64")
+
+oci_image_index(
+    name = "app_image",
+    images = [
+        ":app_linux_arm64",
+        ":app_linux_amd64",
+        ":app_windows_amd64",
+    ]
+)
+
+oci_push(
+    image = ":app_image",
+    repository = "ghcr.io/<OWNER>/image",
+    default_tags = ["6.4.1-rc1"]
+)
+```
```

Comment:
Should mention that you need to `bazel run` to push and that the tool accepts `-r|--repository` and `-t|--tag` flags so you can set the repository & tag at runtime. For tag this is particularly useful since when pushing semver versions users probably want the version to come from the latest tag on the repository and not be hard-coded in the BUILD file.